runtime/lua/vim/lsp/completion.lua

```diff
@@ -129,18 +129,33 @@ end
 --- @param item lsp.CompletionItem
 --- @return string
 local function get_completion_word(item)
-  if item.textEdit ~= nil and item.textEdit.newText ~= nil and item.textEdit.newText ~= '' then
-    if item.insertTextFormat == protocol.InsertTextFormat.PlainText then
-      return item.textEdit.newText
-    else
-      return parse_snippet(item.textEdit.newText)
-    end
-  elseif item.insertText ~= nil and item.insertText ~= '' then
-    if item.insertTextFormat == protocol.InsertTextFormat.PlainText then
-      return item.insertText
-    else
+  if item.insertTextFormat == protocol.InsertTextFormat.Snippet then
+    if item.textEdit then
+      -- Use label instead of text if text has different starting characters.
+      -- label is used as abbr (=displayed), but word is used for filtering
+      -- This is required for things like postfix completion.
+      -- E.g. in lua:
+      --
+      --    local f = {}
+      --    f@|
+      --      ▲
+      --      └─ cursor
+      --
+      --    item.textEdit.newText: table.insert(f, $0)
+      --    label: insert
+      --
+      -- Typing `i` would remove the candidate because newText starts with `t`.
+      local text = item.insertText or item.textEdit.newText
```

Comment:
Shouldn't this just be `item.textEdit.newText`? If I'm reading the specs correctly `textEdit` has priority over `insertText`.

Reply:
`item.textEdit.newText` still gets used if the snippet is accepted. The `insertText` is used because it looks as if this is the better choice for the `word` (which will get prefix matches by `complete()`)

For example, in jdtls you get snippet candidates to implement methods from a base class or interface.

Say you have an `AstVisitor` class with various `visitXY` methods, and you implement one.
If you type `visit|` you'll get completion candidates for all the various methods, where an individual item can look like this:

```
  detail: Override method in 'AstVisitor'
  filterText: @Override\npublic Symbol visitAssignment(Assignment<?> node, ExpressionAnalysisContext context) {\n    ${0:// TODO Auto-generated method stub\n    return super.visitAssignment(node, context);}\n}
  insertText: visitAssignment
  insertTextFormat: 2
  kind: 2
  label: visitAssignment(Assignment<?> node, ExpressionAnalysisContext context) : Symbol
  sortText: 999999163
  textEdit: table: 0x7a70f453e628
    newText: @Override\npublic Symbol visitAssignment(Assignment<?> node, ExpressionAnalysisContext context) {\n    ${0:// TODO Auto-generated method stub\n    return super.visitAssignment(node, context);}\n}
    range: table: 0x7a70f47f5bf8
      end: table: 0x7a70f47f5c40
        character: 12
        line: 457
      start: table: 0x7a70f40a3be0
        character: 8
        line: 457
  textEditText: @Override\npublic Symbol visitAssignment(Assignment<?> node, ExpressionAnalysisContext context) {\n    ${0:// TODO Auto-generated method stub\n    return super.visitAssignment(node, context);}\n}
```

Using `textEdit.newText` as word would mean that the entry would disappear, because `visit|` doesn't share the prefix with `@Override...`

Granted, this might be over-fitting towards how `jdtls` represents things and I expect that we'll get issues about other language servers. But I'm fairly confident that we can keep tweaking these heuristics until we get a good experience cross languages. E.g. another option is to choose whatever has a prefix match (and is the shortest).

Reply:
Another option would be to set `equal = 1` in the items for `complete()`, then it would no longer apply the prefix match and not discard the wrong entries. The downside is that if you don't expand the snippet, you get a less useful `word` added. For example it wouldn't handle newlines well.

It would also mean that we'd have to do manual re-filtering after each character - which means either also invoking a new request, or cache the entries in the `Context`.

Reply:
That sounds reasonable, I'm okay with the current approach then.

> Another option would be to set equal = 1 in the items for complete()

Doesn't that affect fuzzy matching?